R/internal.R

```diff
@@ -1,1474 +1,1501 @@
+        } else if (class(override.se) != "list") {
+            warning("SEs must be provided as a list. Using default SEs.")
+            se <- models[[i]]@se
+        } else if (length(override.se) != length(models)) {
+            warning(paste("Number of SEs provided does not match number of models.", 
+                          "Using default SEs."))
+            se <- models[[i]]@se
+        } else if (length(models[[i]]@se) != length(override.se[[i]])) {
+            warning(paste("Number of SEs provided does not match number of ", 
+                          "coefficients in model ", i, ". Using default SEs.", sep = ""))
+            se <- models[[i]]@se
+        } else if (class(override.se[[i]]) != "numeric") {
+            warning(paste("SEs provided for model", i, 
+                          "are not numeric. Using default SEs."))
+            se <- models[[i]]@se
+        } else {
+            se <- override.se[[i]]
+        }
+        models[[i]]@se <- se
+        
+        # p values
+        if (class(override.pval) != "list" && length(override.pval) == 1 && 
+            override.pval == 0) {
+            pval <- models[[i]]@pvalues
+        } else if (class(override.pval) == "numeric" && length(models) == 1 && 
+                   length(override.pval) == length(models[[i]]@pvalues)) {
+            pval <- override.pval
+        } else if (class(override.pval) != "list") {
+            warning("p values must be provided as a list. Using default p values.")
+            pval <- models[[i]]@pvalues
+        } else if (length(override.pval) != length(models)) {
+            warning(paste("Number of p values provided does not match number of", 
+                          "models. Using default p values."))
+            pval <- models[[i]]@pvalues
+        } else if (length(models[[i]]@se) != length(override.pval[[i]])) {
+            # previous line: comparison with se because pvalues can be empty
+            warning(paste("Number of p values provided does not match number of ", 
+                          "coefficients in model ", i, ". Using default p values.", sep = ""))
+            pval <- models[[i]]@pvalues
+        } else if (class(override.pval[[i]]) != "numeric") {
+            warning(paste("p values provided for model", i, 
+                          "are not numeric. Using default p values."))
+            pval <- models[[i]]@pvalues
+        } else {
+            pval <- override.pval[[i]]
+        }
+        models[[i]]@pvalues <- pval
+        
+        # lower bound of confidence intervals
+        if (is.null(override.ci.low)) {
+            # do nothing
+        } else if (class(override.ci.low) != "list" && length(override.ci.low) 
+                   == 1 && override.ci.low == 0) {
+            ci.low <- models[[i]]@ci.low
+        } else if (class(override.ci.low) == "numeric" && length(models) == 1 && 
+                   length(override.ci.low) == length(models[[i]]@coef)) {
+            ci.low <- 
```

Comment:
Could there be cases where there is a closing parenthesis that is not closing `I`? Perhaps inside the expression? If so, can you make sure only the right closing parenthesis is removed?

Reply:
sure, this should work. I'm explaining how I'm reasoning around it and what I did at the end: 

<<if (grepl("I\\(", rownames(m)[i]) == TRUE)>>  makes it work only for "I" occurrences. 
In case someone uses another round parenthesis inside the "I" expression: 

-the opening one is replaced only if it is next to "I", so any other case is not affected.

- then, using sub instead of gsub, it replaces only one occurrence, so in case there is another closing parenthesis it should stay there.

Reply:
I added *$ to make sure it takes the last parenthesis at the end of the string since a case like this one:
I(disp^(2)+2) 
was giving a wrong output. In fact, maybe there is more stuff after the inner parenthesis and in that case, only using sub wasn't enough. This way should be fine

Reply:
What about cases with multiple `I()` expressions? If only one closing bracket is removed but multiple opening brackets, wouldn't that be a problem? Also, what about cases where there is some text after the expression, in which case `$` wouldn't work? Can you use some examples like these to test it?

Reply:
One possibility may in fact be to use forward- and backward-looking regular expressions. For example, first identify and delete `)` that comes after `I(.+` (where `.+` stands for one or more arbitrary characters), then delete `I(`. Google "regex negative lookbehind" and try it out with some online regex test tool.

Reply:
correct me if I am wrong. If you use I(...) in one regression variable the expression relative to that term is inside I(...).  You can use another I(...) for another variable, but not the same one. Is this correct? 
If there is some text after the expression that works, since it is taking away the last ) nothing else.
I can definitely test it.

Reply:
I don't know, to be honest. If that's the case, my additional comments may be moot. Let me know if you need to test anything else, and if you think we're good to go, I'll merge the PR.

Reply:
I'll check if the "I" works that way (I need to be 100% sure) if it doesn't I'll try the regex negative lookbehind options. I'll keep you posted!

Reply:
Hi Philip, 

I think that what I said above is correct (for more info [https://docs.tibco.com/pub/enterprise-runtime-for-R/3.2.0/doc/html/Language_Reference/base/I.html]

so I am considering only the case where the string finishes by ")".

However, I did more test on the previous version and decided to go for a safer solution. 

This new solution
- does not touch the string inside I(..) so even if somebody puts more parenthesis or math inside doesn't matter
- The problem was with more parentheses at the end. The regex was only reading one, hence solving the problem just for one final parenthesis (the I's one). Hence, I added some if statements to account for four cases: from 1 to 4 parenthesis at the end. I don't think that more than 4 can help. To do so, I had to insert a new function to count the closing parentheses. I placed it before the "outputmatrix" function (lines 617-623). 

Hope you like this one

Reply:
Thanks. It looks better, but is still a bit hacky... How about the pattern `"(.*)(?:I\\()(.+)(?:\\))(.*)"`? This uses capturing groups and non-capturing groups. The first part is a capturing group that records everything before the `I(`. The second part is a non-capturing group for `I(`, meaning that this part is just ignored. The third part is the main part that we want. The fourth part is the closing bracket. And the last part is anything that comes after the closing bracket. Here are some examples:

```
pattern <- "(.*)(?:I\\()(.+)(?:\\))(.*)"

gsub(pattern, "\\1", "testI(disp^2)test2")  # returns "test"
gsub(pattern, "\\2", "testI(disp^2)test2")  # returns "disp^2"
gsub(pattern, "\\3", "testI(disp^2)test2")  # returns "test2"

gsub(pattern, "\\1", "I(disp^2)")  # returns ""
gsub(pattern, "\\2", "I(disp^2)")  # returns "disp^2"
gsub(pattern, "\\3", "I(disp^2)")  # returns ""

gsub(pattern, "\\1", "I(disp^2*(4 + 3))")  # returns ""
gsub(pattern, "\\2", "I(disp^2*(4 + 3))")  # returns "disp^2*(4 + 3)"
gsub(pattern, "\\3", "I(disp^2*(4 + 3))")  # returns ""
```

Let me know if that works...

Reply:
done. Waw! very elegant solution!!!